src/passes/SimplifyLocals.cpp

```diff
@@ -1067,35 +1060,70 @@ struct SimplifyLocals
             }
             return ret;
           };
+
+          // Pick the index with the most uses - maximizing the chance to
+          // lower one's uses to zero. If types differ though then we prefer to
+          // switch to a more refined type even if there are fewer uses, as that
+          // may have significant benefits to later optimizations (we may be
+          // able to use it to remove casts, etc.).
+          auto* func = this->getFunction();
           Index best = -1;
           for (auto index : *set) {
-            if (best == Index(-1) ||
+            if (best == Index(-1)) {
+              // This is the first possible option we've seen.
+              best = index;
+              continue;
+            }
+
+            auto bestType = func->getLocalType(best);
+            auto indexType = func->getLocalType(index);
+            if (!Type::isSubType(indexType, bestType)) {
+              // This is less refined than the current best; ignore.
+              continue;
+            }
+
+            // This is better if it has a more refined type, or if it has more
+            // uses.
+            if (indexType != bestType ||
                 getNumGetsIgnoringCurr(index) > getNumGetsIgnoringCurr(best)) {
               best = index;
             }
           }
           assert(best != Index(-1));
           // Due to ordering, the best index may be different from us but have
           // the same # of locals - make sure we actually improve.
-          if (best != curr->index && getNumGetsIgnoringCurr(best) >
-                                       getNumGetsIgnoringCurr(curr->index)) {
+          auto bestType = func->getLocalType(best);
+          auto oldType = func->getLocalType(curr->index);
+          if (best != curr->index && (getNumGetsIgnoringCurr(best) >
+                                        getNumGetsIgnoringCurr(curr->index) ||
+                                      bestType != oldType)) {
             // Update the get counts.
             (*numLocalGets)[best]++;
             assert((*numLocalGets)[curr->index] >= 1);
             (*numLocalGets)[curr->index]--;
             // Make the change.
             curr->index = best;
             anotherCycle = true;
+            if (bestType != oldType) {
+              curr->type = func->getLocalType(best);
+              // We are switching to a more refined type, which might require
+              // changes in the user of the local.get.
+              refinalize = true;
+            }
           }
         }
       }
     };
 
     EquivalentOptimizer eqOpter;
     eqOpter.module = this->getModule();
+    eqOpter.passOptions = this->getPassOptions();
```

Comment:
Is the `this->` necessary here?

Reply:
Unfortunately yes,
```cpp
SimplifyLocals.cpp:1120:27: error: there are no arguments to ‘getPassOptions’ that depend on a template parameter, so a declaration of ‘getPassOptions’ must be available [-fpermissive]
 1120 |     eqOpter.passOptions = getPassOptions();
      |                           ^~~~~~~~~~~~~~
SimplifyLocals.cpp:1120:27: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
```
Or, from clang,
```cpp
SimplifyLocals.cpp:1120:27: error: use of undeclared identifier 'getPassOptions'; did you mean 'PassOptions'?
    eqOpter.passOptions = getPassOptions();
                          ^
```
Something to do with all this being inside a templated struct, I think.